src/lib.rs

```diff
@@ -96,75 +94,80 @@ impl<'a> MownStr<'a> {
         debug_assert!(self.is_owned(), "extract_box() called on borrowed MownStr");
         // extract data to make box
         let ptr = self.addr.as_ptr();
-        let len = self.real_len();
         // turn to borrowed, to avoid double-free
-        self.xlen = 0;
+        self.owners = None;
         debug_assert!(self.is_borrowed());
         // make box
-        let slice = slice::from_raw_parts_mut(ptr, len);
+        let slice = slice::from_raw_parts_mut(ptr, self.len);
         let raw = str::from_utf8_unchecked_mut(slice) as *mut str;
         Box::from_raw(raw)
     }
 }
 
 impl<'a> Drop for MownStr<'a> {
     fn drop(&mut self) {
-        if self.is_owned() {
+        // if this is the last owner for the underlying string
+        if let Some(1) = self
+            .owners
+            .as_ref()
+            .map(|o| o.fetch_sub(1, Ordering::Relaxed))
+        {
             unsafe {
-                std::mem::drop(self.extract_box());
+                drop(self.extract_box());
             }
         }
     }
 }
 
 impl<'a> Clone for MownStr<'a> {
-    fn clone(&self) -> MownStr<'a> {
-        if self.is_owned() {
-            Box::<str>::from(self.deref()).into()
-        } else {
-            MownStr {
-                addr: self.addr,
-                xlen: self.xlen,
-                _phd: self._phd,
-            }
+    fn clone(&self) -> Self {
+        self.owners
+            .as_ref()
+            .map(|o| o.fetch_add(1, Ordering::Relaxed));
+
+        MownStr {
+            addr: self.addr,
+            len: self.len,
+            _phd: self._phd,
+            owners: self.owners.clone(),
```

Comment:
This radically changes the way `MownStr` was supposed to work. Namely, when cloning an owned MownStr, the text data should be duplicated, just like when cloning a `Box<str>`. That's why, IMO, there should *not* be any double-free even after a clone.

Reply:
@pchampin  Totally agree, the intent was simply to satisfy miri